README
Swaram : A Frontend for Festival
================================

Copyright 2003  Baiju M <baijum81@lycos.com>

Swaram is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

Swaram is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Swaram; see the file COPYING.  If not, write to
the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.
----------------------------------------------------------------------

----------------------------------------------------------------------

Copyright and License of the scheme script 'nt2w' is given in that file.
The original name of that file is 'text2wave', which is available with
'festival' package. The modifications made to the file are :
1. File name changed.
2. The second line of the script changed to :-
   "true" ; exec /usr/bin/festival --script $0 $*

If next version of 'festival' fix this problem, 'nt2w' will be removed from 
this package.
----------------------------------------------------------------------


Installation
-=-=-=-=-=-=

1. Login as 'root'
2. run install.sh script, to run :

	$./install.sh

(You must install 'festival' and 'sox' packages)


Usage
=-=-=

Enter 'swaram' command from terminal.

_________________________________________
_________________________________________

THIS PACKAGE IS ONLY TESTED IN RedHat 9.0 
_________________________________________
_________________________________________



ChangeLog
2003-10-29  Baiju M  <baijum81@lycos.com>

	First public release of Swaram.

	* COPYING, ChangeLog, FAQ, README, TODO, VERSION, install.sh, mainWindow.py, messageDialog.py, nt2w, prefWindow.py, swaram, swaram.conf, swaram.desktop, swaram.glade, swaram.png, swaram.py, swaram.xpm, swaramBase.py:
	New files.


install.sh
#!/bin/sh

VERSION=`cat VERSION`
INS_PATH=/usr

# Checking whether 'festival' is installed or not
which festival > /dev/null 2>&1
if [ $? == 0 ];then
	echo -n ""
else
	echo "Please install 'festival' package, otherwise Swaram will not work."
fi

# Checking whether 'sox' is installed or not
which play > /dev/null 2>&1
if [ $? == 0 ];then
	echo -n ""
else
	echo "Please install 'sox' package, otherwise Swaram will not work."
fi

# Checking whether 'xpdf' is installed or not
which pdftotext > /dev/null 2>&1
if [ $? == 0 ];then
	echo -n ""
else
	echo "Please install 'xpdf' package, otherwise PDF files cannot play."
fi

# Checking whether 'lynx' is installed or not
which lynx > /dev/null 2>&1
if [ $? == 0 ];then
	echo -n ""
else
	echo "Please install 'lynx' package, otherwise HTML files cannot play."
fi

# Installing the basic stuff
mkdir -p $INS_PATH/share/swaram
cp -f swaram swaram.glade swaram.py swaramBase.py mainWindow.py prefWindow.py messageDialog.py swaram.xpm nt2w $INS_PATH/share/swaram
ln -sf $INS_PATH/share/swaram/swaram $INS_PATH/bin/swaram
ln -sf $INS_PATH/share/swaram/nt2w $INS_PATH/bin/nt2w

# Installing swaram.desktop , swaram.png and swaram.conf
cp -f swaram.desktop $INS_PATH/share/applications/swaram.desktop
cp -f swaram.png $INS_PATH/share/pixmaps/swaram.png
cp -f swaram.conf /etc/swaram.conf

# Installing doc files
mkdir -p $INS_PATH/share/doc/swaram-$VERSION
cp -f COPYING FAQ README ChangeLog TODO $INS_PATH/share/doc/swaram-$VERSION


prefWindow.py
#!/usr/bin/env python

# mainWindow - preference window interface for Swaram
# Copyright (C) 2003 Baiju M
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
# Author: Baiju M <baijum81@lycos.com>

import os
import string
import gtk

import mainWindow
import swaramBase

class prefWindow:
	def __init__(self, parent, xml):
		self.parent = parent

		self.prefWin = xml.get_widget('prefWindow')
 		self.prefWin.connect("delete-event", self.on_pref_close_button_clicked)
		self.prefWin.set_position(gtk.WIN_POS_CENTER)
		self.pref_cbutton=xml.get_widget('pref_cbutton')
		self.pref_close_button = xml.get_widget('pref_close_button')
		self.table1 = xml.get_widget('table1')
		self.textViewer = xml.get_widget('textViewerEntry')
		self.htmlViewer = xml.get_widget('htmlViewerEntry')
		self.pdfViewer = xml.get_widget('pdfViewerEntry')

		xml.signal_connect("on_pref_close_button_clicked", self.on_pref_close_button_clicked)
		xml.signal_connect("on_pref_cbutton_toggled", self.on_pref_cbutton_toggled)

	def busy(self):
		self.prefWin.set_sensitive(gtk.FALSE)
		self.prefWin.window.set_cursor(busy_cursor)

	def newPrefWin(self):
		swaramBaseConfigure = swaramBase.swaramBase()
		(text_viewer, html_viewer, pdf_viewer, viewer_state) = swaramBaseConfigure.get_viewers()
		self.textViewer.set_text(text_viewer)
		self.htmlViewer.set_text(html_viewer)
		self.pdfViewer.set_text(pdf_viewer)
		if viewer_state == 'enable':
			self.pref_cbutton.set_active(gtk.TRUE)
		else:
			self.pref_cbutton.set_active(gtk.FALSE)
		self.prefWin.show_all()

	def on_pref_close_button_clicked(self, *args):
		swaramBaseConfigure = swaramBase.swaramBase()

		textViewer = self.textViewer.get_text()
		htmlViewer = self.htmlViewer.get_text()
		pdfViewer = self.pdfViewer.get_text()

		if self.pref_cbutton.get_active() == gtk.TRUE:
			viewerState = 'enable'
		else:
			viewerState = 'disable'

		swaramBaseConfigure.set_viewers(textViewer,htmlViewer,pdfViewer,viewerState)

		self.prefWin.hide()
		return gtk.TRUE
	
	def on_pref_cbutton_toggled(self, *args):
		self.table1.set_sensitive(self.pref_cbutton.get_active())


swaramBase.py
#!/usr/bin/env python

# swaramBase - basic functions for Swaram
# Copyright (C) 2003 Baiju M
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
# Author: Baiju M <baijum81@lycos.com>

import time
import os
import string
import gtk
import messageDialog

text_viewer = 'less'
html_viewer = 'lynx'
pdf_viewer = 'gv'
viewer_state = 'enable'

class swaramBase:

    def __init__(self):
        pass

    def setup_conf_files(self):
        global text_viewer
        global html_viewer
        global pdf_viewer
        global viewer_state

        if os.access('/etc/swaram.conf',os.R_OK):

            pf = open('/etc/swaram.conf', 'r')
            prefs = pf.readlines()
            pf.close()

            try:

                for line in prefs:
                    line = string.strip(line)
                    if line != '' and line[0] != '#':
                        preference, values = string.split(line, "=")
                        preference = string.strip(preference)
                        if preference == "text_viewer":
                            text_viewer = string.strip(values)
                        elif preference == "html_viewer":
                            html_viewer = string.strip(values)
                        elif preference == "pdf_viewer":
                            pdf_viewer = string.strip(values)
                        elif preference == "viewer_state":
                            viewer_state = string.strip(values)

            except:
                messageDialog.show_message_dialog("/etc/swaram.conf file correpted.")
                pass
        else:
            messageDialog.show_message_dialog("/etc/swaram.conf is not available for reading.")
            pass

        self.read_swaramrc()

    def get_text_viewer(self):
        global text_viewer
        self.read_swaramrc()
        return text_viewer

    def get_html_viewer(self):
        global html_viewer
        self.read_swaramrc()
        return html_viewer

    def get_pdf_viewer(self):
        global pdf_viewer
        self.read_swaramrc()
        return pdf_viewer

    def get_viewer_state(self):
        global viewer_state
        self.read_swaramrc()
        return viewer_state

    def get_viewers(self):
        global text_viewer, html_viewer, pdf_viewer, viewer_state
        self.read_swaramrc()
        return (text_viewer, html_viewer, pdf_viewer, viewer_state)

    def set_viewers(self, textViewer='nil', htmlViewer='nil', pdfViewer='nil', viewerState='enable'):
        global text_viewer
        global html_viewer
        global pdf_viewer
        global viewer_state

        text_viewer = textViewer
        html_viewer = htmlViewer
        pdf_viewer = pdfViewer
        viewer_state = viewerState

        settings = os.environ.get('HOME')+'/.swaramrc'

        try:
            pf = open(settings, 'r')
            self.prefs = pf.readlines()
            pf.close()

            fd = open(settings, 'w')

            for line in self.prefs:
                if line == '' or line[0] == '#':
                    fd.write(line)
                else:
                    tokens = string.split(line, '=')
                    if tokens[0] == 'text_viewer':
                        fd.write("text_viewer="+text_viewer+"\n")
                    elif tokens[0] == 'html_viewer':
                        fd.write("html_viewer="+html_viewer+"\n")
                    elif tokens[0] == 'pdf_viewer':
                        fd.write("pdf_viewer="+pdf_viewer+"\n")
                    elif tokens[0] == 'viewer_state':
                        fd.write("viewer_state="+viewer_state+"\n")
            fd.close()

        except:
            pass

    def read_swaramrc(self):

        global text_viewer
        global html_viewer
        global pdf_viewer
        global viewer_state

        swaramrc = os.environ.get('HOME')+'/.swaramrc'

        try:
            pf = open(swaramrc, 'r')
            prefs = pf.readlines()
            pf.close()
        except:
            fd = open(swaramrc,'w')
            fd.write("text_viewer=less\n")
            fd.write("html_viewer=lynx\n")
            fd.write("pdf_viewer=gv\n")
            fd.write("viewer_state=enable\n")
            fd.close()

        pf = open(swaramrc, 'r')
        prefs = pf.readlines()
        pf.close()

        try:

            for line in prefs:
                line = string.strip(line)
                if line != '' and line[0] != '#':
                    preference, values = string.split(line, "=")
                    preference = string.strip(preference)
                    if preference == "text_viewer":
                        text_viewer = string.strip(values)
                    elif preference == "html_viewer":
                        html_viewer = string.strip(values)
                    elif preference == "pdf_viewer":
                        pdf_viewer = string.strip(values)
                    elif preference == "viewer_state":
                        viewer_state = string.strip(values)

        except:
            fd = open(swaramrc,'w')
            fd.write("text_viewer=less\n")
            fd.write("html_viewer=lynx\n")
            fd.write("pdf_viewer=gv")
            fd.write("viewer_state=enable\n")
            fd.close()

    def read(self, tmp_time, no_of_lines):
        fdp1 = os.popen4("ls x* > tmp_files;for i in `cat tmp_files`;do nt2w -o $i.wav $i;done")

        fd = open('/tmp/'+tmp_time+'/xaaa','r')
        lines=fd.readlines()
        fd.close()
        
        total_chars = 0
        for line in lines:
            total_chars = len(line)+total_chars

	if total_chars < (((no_of_lines - 1)/2)*75):
            time.sleep(no_of_lines*5)

        i = 0
        while i < 61:
            time.sleep(5)
            if os.access('/tmp/'+tmp_time+'/xaaa.wav', os.F_OK):
                fdp2 = os.popen4("for i in `cat tmp_files`;do play $i.wav;done")
                break
            elif i == 60:
                print "Sorry! Your system is too slow"
                break
            else:
                i = i + 1

    def split_lines(self,length):
        for i in range(5,50,2):
            if length < ((26**3)*i):
		os.system('split --lines='+str(i)+' --suffix-length=3 new_file.txt')
                return i
                break

    def open_viewer(self, viewer, file, man='no', info='no'):
        if viewer == 'nil':
            pass
        elif viewer == "less":
            os.spawnlp(os.P_NOWAIT, 'xterm','xterm','-e','less',file)
        elif viewer == "lynx":
            os.spawnlp(os.P_NOWAIT, 'xterm','xterm','-e','lynx',file)
        elif viewer == "links":
            os.spawnlp(os.P_NOWAIT, 'xterm','xterm','-e','links',file)
        else:
            try:
                os.spawnlp(os.P_NOWAIT, viewer, viewer, file)
            except:
                messageDialog.show_message_dialog(viewer+" is not supported.")
                pass


mainWindow.py
#!/usr/bin/env python

# mainWindow - main interface window for Swaram
# Copyright (C) 2003 Baiju M
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
# Author: Baiju M <baijum81@lycos.com>

import signal
import gtk
import gobject
import gtk.glade
import gnome.ui
import string
import sys
import time
import os
import re

import swaramBase
import prefWindow
import messageDialog

swaramBaseConfigure = swaramBase.swaramBase()
swaramBaseConfigure.setup_conf_files()

home = os.environ.get('HOME')
if not os.access(home+'/.swaram', os.F_OK):
    os.mkdir(home+'/.swaram',0755)
swaram_dir = home+'/.swaram'
tmp_time = time.ctime()
man_state = 0

domain = "swaram"

if os.access("swaram.glade", os.F_OK):
    xml = gtk.glade.XML ("./swaram.glade", domain=domain)
else:
    xml = gtk.glade.XML ("/usr/share/swaram/swaram.glade", domain=domain)

if os.access("VERSION", os.F_OK):
    version_file = "./VERSION"
else:
    version_file = '/usr/share/swaram/VERSION'

fd = open(version_file,'r')
VERSION = fd.readline()
fd.close()

class mainWindow:

    def version():
        global VERSION
        return VERSION
    
    def destroy(self, *args):
        self.on_file_stop_button_clicked()
        gtk.mainquit()

    def __init__(self):
        self.toplevel = xml.get_widget ('mainWindow')
        self.notebook = xml.get_widget('notebook1')
        self.text_read_button = xml.get_widget("text_read_button")
        self.text_close_button = xml.get_widget("text_close_button")
        self.file_read_button = xml.get_widget("file_read_button")
        self.file_save_button = xml.get_widget("file_save_button")
        self.file_pause_button = xml.get_widget("file_pause_button")
        self.file_stop_button = xml.get_widget("file_stop_button")
        self.file_close_button = xml.get_widget("file_close_button")
        self.file_about_button = xml.get_widget("file_about_button")
        self.man_read_button = xml.get_widget("man_read_button")
        self.man_stop_button = xml.get_widget("man_stop_button")
        self.man_close_button = xml.get_widget("man_close_button")
        self.man_about_button = xml.get_widget("man_about_button")
        self.textEntryToRead = xml.get_widget('textEntryToRead')
        self.fileEntryToRead = xml.get_widget('fileEntryToRead')
        self.manEntryToRead = xml.get_widget('manEntryToRead')

        self.toplevel.connect("destroy", self.destroy)

        self.prefWin = prefWindow.prefWindow(self,xml)

        self.toplevel.show_all ()

        xml.signal_connect("on_text_read_button_clicked", self.on_text_read_button_clicked)
        xml.signal_connect("on_text_close_button_clicked", self.on_text_close_button_clicked)
        xml.signal_connect("on_file_read_button_clicked", self.on_file_read_button_clicked)
        xml.signal_connect("on_file_save_button_clicked", self.on_file_save_button_clicked)
        xml.signal_connect("on_file_pause_button_clicked", self.on_file_pause_button_clicked)
        xml.signal_connect("on_file_stop_button_clicked", self.on_file_stop_button_clicked)
        xml.signal_connect("on_file_close_button_clicked", self.on_text_close_button_clicked)
        xml.signal_connect("on_file_about_button_clicked", self.on_file_about_button_clicked)
        xml.signal_connect("on_file_pref_button_clicked", self.on_file_pref_button_clicked)
        xml.signal_connect("on_man_read_button_clicked", self.on_man_read_button_clicked)
        xml.signal_connect("on_man_stop_button_clicked", self.on_file_stop_button_clicked)
        xml.signal_connect("on_man_close_button_clicked", self.on_text_close_button_clicked)
        xml.signal_connect("on_man_about_button_clicked", self.on_file_about_button_clicked)
        xml.signal_connect("on_man_pref_button_clicked", self.on_file_pref_button_clicked)
        xml.signal_connect("on_man_rbutton_toggled", self.on_manual_toggled)
        xml.signal_connect("on_info_rbutton_toggled", self.on_manual_toggled)
        xml.signal_connect("on_exit_activate", self.on_exit_activate)

        gtk.mainloop()

    def on_text_read_button_clicked(self, *args):
        self.on_file_stop_button_clicked()
        text = self.textEntryToRead.get_text()
        print "You are asked to read :" , text
        f=open(swaram_dir+'/gf','w')
        f.write(text)
        f.close()
        os.system('nt2w -o '+swaram_dir+'/gf.wav '+swaram_dir+'/gf')
        os.system('play '+swaram_dir+'/gf.wav')

    def on_text_close_button_clicked(self, *args):
        self.on_file_stop_button_clicked()
        gtk.mainquit()

    def on_file_read_button_clicked(self, *args):
        self.on_file_stop_button_clicked()
        global tmp_time

        file = self.fileEntryToRead.get_text()
        if os.access('/tmp/',3) == 0 :
            messageDialog.show_message_dialog("Sorry! you don't have write access to /tmp/ directory")
            gtk.mainquit()
        print "You are asked to read :" , file
        tmp_time = time.ctime()

        os.mkdir('/tmp/'+tmp_time,0744)
        os.chdir('/tmp/'+tmp_time)

        fto = 'fto'
        ftt = 'ftt'

        try:
            file_type_one = re.search('(?<=\.)htm', file)
            fto = file_type_one.group(0)
        except:
            pass

        try:
            file_type_two = re.search('(?<=\.)pdf', file)
            ftt = file_type_two.group(0)
        except:
            pass

        viewer_state = swaramBaseConfigure.get_viewer_state()

        try:
            if fto == 'htm':
                viewer = swaramBaseConfigure.get_html_viewer()
                if viewer_state == 'disable':
                    viewer = 'nil'
                os.spawnlp(os.P_WAIT, 'cp', 'cp', file, '/tmp/'+tmp_time+'/new_file.html')
                fd = os.popen('lynx -nolist -dump new_file.html > new_file.txt.tmp')
                fd.close()
            elif ftt == 'pdf':
                viewer = swaramBaseConfigure.get_pdf_viewer()
                if viewer_state == 'disable':
                    viewer = 'nil'
                os.spawnlp(os.P_WAIT, 'cp', 'cp', file, '/tmp/'+tmp_time+'/new_file.pdf')
                ##FIXME: .tmp extension, any problem?
                os.spawnlp(os.P_WAIT, 'pdftotext', 'pdftotext', 'new_file.pdf', 'new_file.txt.tmp')
            else:
                viewer = swaramBaseConfigure.get_text_viewer()
                if viewer_state == 'disable':
                    viewer = 'nil'
                os.spawnlp(os.P_WAIT, 'cp', 'cp', file, '/tmp/'+tmp_time+'/new_file.txt.tmp')
        except:
            pass

        fd = os.popen('fmt new_file.txt.tmp > new_file.txt.tmp2')
        fd.close()
        swaramBaseConfigure.open_viewer(viewer, file)
        self.read_text_file()

    def on_file_save_button_clicked(self, *args):
        print "file_save_button_clicked"

    def on_file_pause_button_clicked(self, *args):
        print "file_pause_button_clicked"

    def on_file_stop_button_clicked(self, *args):
        global tmp_time
        os.spawnlp(os.P_WAIT, 'killall', 'killall', '-q9', 'festival')
        os.spawnlp(os.P_WAIT, 'killall', 'killall', '-q9', 'sox')
        os.spawnlp(os.P_WAIT, 'rm', 'rm', '-rf', '/tmp/'+tmp_time)

    def on_file_about_button_clicked(self, *args):
        global VERSION
        dlg = gnome.ui.About("Swaram",
                             VERSION,
                             "Copyright 2003 Baiju M <baijum81@lycos.com>",
                             "Swaram is a frontend for 'festival'\n"
                             "Swaram is licensed under GNU GPL\n"
                             "Please see the COPYING file with the source",
                             ())
        dlg.run()
	dlg.destroy()

    def on_man_read_button_clicked(self, *args):
        self.on_file_stop_button_clicked()
        global tmp_time,man_state

        viewer_state = swaramBaseConfigure.get_viewer_state()
	viewer = swaramBaseConfigure.get_text_viewer()
        if viewer_state == 'disable':
            viewer = 'nil'

        man  = self.manEntryToRead.get_text()
        if os.access('/tmp/',3) == 0 :
            messageDialog.show_message_dialog("Sorry! you don't have write access to /tmp/ directory")
            gtk.mainquit()

        print "You are asked to read :" , man
        tmp_time = time.ctime()
        os.mkdir('/tmp/'+tmp_time,0744)
        os.chdir('/tmp/'+tmp_time)
        if man_state == 1:
            fd = os.popen('info '+str(man)+' > new_file.txt.tmp')
            fd.close()
            swaramBaseConfigure.open_viewer(viewer,'new_file.txt.tmp')
            self.read_text_file()
        elif man_state == 0:
            fd = os.popen('man '+str(man)+' | col -bx > new_file.txt.tmp')
            fd.close()
            swaramBaseConfigure.open_viewer(viewer,'new_file.txt.tmp')
            self.read_text_file()

    def on_manual_toggled(self, *args):
        global man_state
        if xml.get_widget ('man_rbutton').get_active():
            man_state = 0 
        elif xml.get_widget ('info_rbutton').get_active():
            man_state = 1

    def on_file_pref_button_clicked(self, *args):
        self.prefWin.newPrefWin()

    def on_exit_activate(self, *args):
        self.destroy()

    def read_text_file(self, *args):
        global tmp_time
        fd = os.popen('fmt new_file.txt.tmp > new_file.txt.tmp2')
        fd.close()
        fd = open('new_file.txt.tmp2','r')
        lines = fd.readlines()
        fd.close()

        fd=open('new_file.txt','w')
        for i in range(len(lines)):
            if lines[i] != '\n':
                if lines[i-1] == '\n' and i != 0:
                    fd.write('\n'+lines[i])
                else:
                    fd.write(lines[i])
        fd.close()

        fd = open('new_file.txt','r')
        lines = fd.readlines()
        length = len(lines)
        fd.close()

        no_of_lines = swaramBaseConfigure.split_lines(length)

        swaramBaseConfigure.read(tmp_time, no_of_lines)
